File: intervals/insert_interval.js
var insert = function (intervals, newInterval) {
  intervals.push(newInterval);
  // Sorting by start time
  intervals.sort((a, b) => a[0] - b[0]);

  const merged = [];
  let start = intervals[0][0];
  let end = intervals[0][1];

  for (let i = 1; i < intervals.length; i++) {
    const interval = intervals[i];

    // Overlapping intervals, determine the higher end time
    if (interval[0] <= end) end = Math.max(end, interval[1]);
    // Non-overlapping intervals, we push a new interval in
    else {
      merged.push([start, end]);
      start = interval[0];
      end = interval[1];
    }
  }

  // Adding the last interval
  merged.push([start, end]);
  return merged;
};

File: intervals/insert_interval.py
class Solution:
    def insert(self, intervals: List[List[int]], new_interval: List[int]) -> List[List[int]]:
        i = 0
        merged = []

        # Find all intervals that come before new_interval
        while i < len(intervals) and intervals[i][1] < new_interval[0]:
            merged.append(intervals[i])
            i += 1

        # Merge all intervals that overlap with 'new_interval'
        while i < len(intervals) and new_interval[1] >= intervals[i][0]:
            new_interval[0] = min(new_interval[0], intervals[i][0])
            new_interval[1] = max(new_interval[1], intervals[i][1])
            i += 1

        # Append new_interval
        merged.append(new_interval)

        # Add the remaining intervals
        while i < len(intervals):
            merged.append(intervals[i])
            i += 1

        return merged

# Time Complexity: O(n)
# Space Complexity: O(1)